File: frontend/src/app/services/habit.service.ts
import { Injectable } from '@angular/core';
import { Habit, HabitProgress } from '../models/habit';
import { HttpClient } from '@angular/common/http';
import { map, Observable } from 'rxjs';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class HabitService {

  private url : string = environment.apiUrl + "/api/habits";

  constructor(
    private http : HttpClient
  ) { }

  findAll() : Observable<Habit[]>{
    return this.http.get<Habit[]>(this.url)
  }

  findByUserId(id : number) : Observable<Habit[]>{
    return this.http.get<Habit[]>(`${this.url}/user/${id}`)
  }

  findById(id : number) : Observable<Habit>{
    return this.http.get<Habit>(`${this.url}/${id}`)
  }

  create(habit : Habit) : Observable<Habit>{
    return this.http.post<Habit>(this.url, habit)
  }

  update(habit : Habit) : Observable<Habit>{
    return this.http.put<Habit>(`${this.url}/${habit.id}`, habit)
  }

  updateBatch(habits : Habit[]) : Observable<Habit[]>{
    return this.http.put<Habit[]>(`${this.url}/batch`, habits)
  }

  delete(id : number) : Observable<number>{
    return this.http.delete<number>(`${this.url}/${id}`)
    .pipe(
      map( () => id)
    );
  }
}

File: backend/appbito/src/main/java/com/appbito/appbito/services/HabitService.java
package com.appbito.appbito.services;

import java.util.List;
import java.util.Optional;

import com.appbito.appbito.entities.HabitDTO;

public interface HabitService {
    List<HabitDTO> findAll();
    Optional<HabitDTO> findById(Long id);
    HabitDTO save(HabitDTO habit);
    Optional<HabitDTO> update(HabitDTO habit, Long id);
    void deleteById(Long id);
    
    List<HabitDTO> findByToday();
    List<HabitDTO> findByUserId(Long id);  
    List<HabitDTO> updateBatch(List<HabitDTO> habits);

    List<HabitDTO> migrateProgressData();
}
